Change UserAgent App WindowsPowerShell -> PowerShell

markekraus opened the pull request:
fixes #4911

* renames the App portion of the user agent to `PowerShell`
* Adjusts tests.

markekraus committed:
Change UserAgent App WindowsPowerShell -> PowerShell

msftclas commented:
@markekraus,
Thanks for having already signed the Contribution License Agreement. Your agreement was validated by Microsoft. We will now review your pull request.
_Thanks,
Microsoft Pull Request Bot_

markekraus committed:
[Feature] Run Feature tests

SteveL-MSFT reviewed (approved):
Thanks!

adityapatwardhan commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -720,7 +720,7 @@ Describe "Invoke-WebRequest tests" -Tags "Feature" {
         $jsonContent = $result.Output | ConvertFrom-Json
         $jsonContent.headers.Host | Should Be $uri.Authority
-        $jsonContent.headers.'User-Agent' | Should Match "WindowsPowerShell"
+        $jsonContent.headers.'User-Agent' | Should Match "PowerShell"
```
This match would work with `WindowsPowerShell` as well. This should be `^PowerShell`. 

In addition to that, it should be `MatchExactly`

iSazonov replied:
User-Agent can be `PowerShell/6.0.0-Beta.8` or `PowerShell/6.0.0`- so if we use  MatchExactly the pattern should be "^PowerShell/\d+\.\d+\.\d+.*"

adityapatwardhan replied:
Makes sense.

markekraus replied:
It's actually more like `Mozilla/5.0 (Windows NT; Microsoft Windows 10.0.15063 ; en-US) WindowsPowerShell/6.0.0` and it will be more dynamic "soon"

we could pull the exact one through reflection in `BeforeAll`. Honestly, I'm not even sure this is a good test. `Should Be $uri.Authority` should be sufficient, IMO.

markekraus replied:
fixed.

adityapatwardhan commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -1322,7 +1322,7 @@ Describe "Invoke-RestMethod tests" -Tags "Feature" {
 
         # Validate response
-        $result.Output.headers.'User-Agent' | Should Match "WindowsPowerShell"
+        $result.Output.headers.'User-Agent' | Should Match "PowerShell"
```
Same as above.

adityapatwardhan commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -1334,7 +1334,7 @@ Describe "Invoke-RestMethod tests" -Tags "Feature" {
         # Validate response
         $result.Output.headers.Host | Should Be $Uri.Authority
-        $result.Output.headers.'User-Agent' | Should Match "WindowsPowerShell"
+        $result.Output.headers.'User-Agent' | Should Match "PowerShell"
```
Same as above

adityapatwardhan commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -1347,7 +1347,7 @@ Describe "Invoke-RestMethod tests" -Tags "Feature" {
         # Validate response
         $result.Output.headers.Host | Should Be $uri.Authority
-        $result.Output.headers.'User-Agent' | Should Match "WindowsPowerShell"
+        $result.Output.headers.'User-Agent' | Should Match "PowerShell"
```
Same as above

adityapatwardhan commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -1360,7 +1360,7 @@ Describe "Invoke-RestMethod tests" -Tags "Feature" {
         # Validate response
         $result.Output.headers.Host | Should Match $uri.Authority
-        $result.Output.headers.'User-Agent' | Should Match "WindowsPowerShell"
+        $result.Output.headers.'User-Agent' | Should Match "PowerShell"
```
Same as above

adityapatwardhan commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -1435,7 +1435,7 @@ Describe "Invoke-RestMethod tests" -Tags "Feature" {
 
         # Validate response
-        $result.Output.headers.'User-Agent' | Should Match "WindowsPowerShell"
+        $result.Output.headers.'User-Agent' | Should Match "PowerShell"
```
Same as above

adityapatwardhan commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -1497,7 +1497,7 @@ Describe "Invoke-RestMethod tests" -Tags "Feature" {
                 # Validate response
                 $result.Output.url | Should Match $uri
-                $result.Output.headers.'User-Agent' | Should Match "WindowsPowerShell"
+                $result.Output.headers.'User-Agent' | Should Match "PowerShell"
```
Same as above

adityapatwardhan commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -1525,7 +1525,7 @@ Describe "Invoke-RestMethod tests" -Tags "Feature" {
         # Validate response
         $result.Output.url | Should Match $uri
-        $result.Output.headers.'User-Agent' | Should Match "WindowsPowerShell"
+        $result.Output.headers.'User-Agent' | Should Match "PowerShell"
```
Same as above

adityapatwardhan commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -1560,7 +1560,7 @@ Describe "Invoke-RestMethod tests" -Tags "Feature" {
         $jsonContent = $result.Output | ConvertFrom-Json
         $jsonContent.headers.Host | Should Be $uri.Authority
-        $jsonContent.headers.'User-Agent' | Should Match "WindowsPowerShell"
+        $jsonContent.headers.'User-Agent' | Should Match "PowerShell"
```
Same as above

markekraus committed:
[feature] Address PR Feedback with more precise assertion

markekraus commented:
Test fails due to #4919

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -421,6 +421,9 @@ $PendingCertificateTest = $false
 if ( test-path /etc/centos-release ) { $PendingCertificateTest = $true }
 
+# Get the default UserAgent through reflection
+$DefaultUserAgent = [Microsoft.PowerShell.Commands.PSUserAgent].GetProperty('UserAgent',@('Static','NonPublic')).GetValue($null,$null)
```
It is dummy test: if C# code will be changed the test will be automatically passed.  I believe we should use explicit desired pattern.
Also please move to `BeforeAll`.

markekraus replied:
That's not the purpose of the tests. They are just testing the response from the server to ensure that expected data sent is also returned. if we want to test the patterns in the UserAgent, that would be a separate set of tests for the `PSUserAgent` class.

I will add a separate describe block for `PSUserAgent` tests and a test for the `App` member (there will be other tests possibly for #4193). I will create a separate issue to add and track adding other tests for `PSUserAgent`.

iSazonov reviewed (commented):
Leave a comment

markekraus committed:
[feature] Address PR Feedback and add PSUserAgent Test

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -2188,3 +2189,10 @@ Describe "Web cmdlets tests using the cmdlet's aliases" -Tags "CI" {
+Describe "PSUserAgent Tests" {
+    It "App Should Match ^PowerShell/\d+\.\d+\.\d+.*" {
+        $app = [Microsoft.PowerShell.Commands.PSUserAgent].GetProperty('App',@('Static','NonPublic')).GetValue($null,$null) 
+        $app | Should Match '^PowerShell/\d+\.\d+\.\d+.*'
```
Here I don't see the point of checking internal property - we are free to change (add/remove) internals in any time. We should test public API. Here "public API" is User-Agent which we send to a server and get it back. We should test the User-Agent format if we want to control the User-Agent. So it is good to use regex pattern. I think that check User-Agent 14 times it's redundant - we need to have one check for each web cmdlet.

markekraus replied:
Again, the point of most of those tests is not to test the User-Agent that is sent. Most of those assertions are being used to ensure meaningful data is returned in a response after a set of parameters has been used to make web requests. The only reason I updated them at all was because the change from `WindowsPowerShell` to `PowerShell` would have caused all these tests to false fail.

I will change the two tests that are explicitly about the User-Agent.

iSazonov replied:
As you mentioned `$uri.Authority` is sufficient. So I'd remove `$DefaultUserAgent` at all. I unlike it as you see 😄

markekraus replied:
I'm ok with that. This entire file needs to be more deeply analyzed and cleaned up. Maybe after I'm done making radical changes to it every day 😆.

markekraus committed:
[feature] add tag to describe block

markekraus committed:
[feature] Remove describe block Add pattern to User-Agent specific tests

markekraus committed:
[feature] Address PF Feedback

Remove all assertions for User-Agent that are not specifically about the User-Agent header.

JamesWTruher commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -463,7 +463,7 @@ Describe "Invoke-WebRequest tests" -Tags "Feature" {
         # Validate response content
         $jsonContent = $result.Output.Content | ConvertFrom-Json
-        $jsonContent.headers.'User-Agent' | Should Match "WindowsPowerShell"
+        $jsonContent.headers.'User-Agent' | Should Match '(?<!Windows)PowerShell\/\d+\.\d+\.\d+.*'
```
I think you might use `MatchExactly` which does a case sensitive match (`-cmatch`). Especially since you're explicitly not matching WindowsPowerShell

markekraus replied:
fixed

markekraus committed:
[feature] Match -> MatchExactly

adityapatwardhan reviewed (approved):
LGTM
